### DISCUSSION TITLE
Run custom FFmpeg media pipelines

### DISCUSSION THREAD

**Discussion by chris-kuhr** [aiortc/aiortc#753] [Q&A] [answered]
Hi *,

I looked into this project for a longer time now and experimented with several things. First of all, I really like this project, good work!

I am trying to adjust the server example for our means, i.e. using a custom FFmpeg installation in combination with other media frameworks for live streaming. We need the received audio data be sent to Jack Audio and the received video data to be retransmitted as UDP multicast stream in MPEG-TS format. I did not find an aiortc-only solution, thus I wanted to only use the signaling part and run our own FFmpeg pipeline via BASH calls.

I tried to do this but failed so far, since I cannot find the right way to deactivate listening to the RTP/UDP transport ports. I store the remote SDP offer in a file and read it with FFmpeg. However, the socket that the aiortc listens to seems to block the FFmpeg pipeline I launched with Popen, because it times out listening to the RTP Ports. I can see via netstat that the negotiated UDP port is already listened to. I already removed all players and recorders, but no luck.

Do you have any suggestions or advise on how to proceed?

Best 
Ck

**Comment by chris-kuhr** [comment]
I solved the first part. I implemented a Jack client thread in a subclass of MediaStreamTrack and added it to a MediaRelay.

But I still have no idea how to resend the incoming video stream as MPEG TS UDP multicast without de- and reencoding.

Any directions are very welcome!

Best 
Ck

**Reply by chris-kuhr** [reply]
I have now tried to use v4l2loopback, but I ran in the same problem described into this [issue](https://github.com/aiortc/aiortc/pull/185)...

**Comment by chris-kuhr** [comment, marked as answer]
Custom code can simply be integrated via deriving MediaStreamTrack.

**Reply by nurik12-py** [reply]
Can you show an example?

**Reply by chris-kuhr** [reply]
Have a look at: [Server.py](https://github.com/aiortc/aiortc/blob/main/examples/server/server.py)
Lines 23-90 show you how its done.
Any specifics are depending on your use case.
You can do this with audio tracks as well.
